migrations/migration.py

```diff
@@ -0,0 +1,178 @@
+        """Generate a new migration script."""
+        current_id = str(self.get_latest_applied_migration_from_fhir())
+        new_id = str(uuid.uuid4())  # Random string as the migration identifier
+        migration_filename = f"{new_id}.py"
+        migration_path = os.path.join(self.migrations_dir, migration_filename)
+        with open(migration_path, "w") as migration_file:
+            migration_file.write(f"# Migration script generated for {migration_name}\n")
+            migration_file.write(f"# Current version: {new_id}\n")
+            migration_file.write(f"# Previous version: {current_id}\n")
+            migration_file.write("\n")
+            migration_file.write("# Add your migration code here\n")
+        return migration_filename
+
+    def run_migrations(self, direction: str):
+        """Run migrations based on the specified direction ("upgrade" or "downgrade")."""
+        if direction not in ["upgrade", "downgrade"]:
+            raise ValueError("Invalid migration direction. Use 'upgrade' or 'downgrade'.")
+
+        current_migration = self.get_latest_applied_migration_from_fhir()
+        applied_migration = None
+
+        if direction == "upgrade":
+            applied_migration = self.get_next_migration(current_migration)
+        elif direction == "downgrade" and current_migration is not None:
+            applied_migration = self.get_previous_migration(current_migration)
+
+        if applied_migration is None:
+            raise ValueError("No valid migration files to run.")
+
+        migration_path = os.path.join(self.migrations_dir, applied_migration)
+        with open(migration_path, "r") as migration_file:
+            migration_code = migration_file.read()
+            try:
+                exec(migration_code)
+                self.update_latest_applied_migration_in_fhir(applied_migration)
+
+            except Exception as e:
+                print(f"Error executing migration {applied_migration}: {e}")
+
+    def get_next_migration(self, current_migration) -> str:
+        """Retrieve the latest."""
+        for current_node, previous_node in self.migration_sequence.items():
+            if previous_node == current_migration:
+                return current_node
+        return None
+
+    def get_previous_migration(self, current_migration) -> str:
+        """Retrieve the latest."""
+        if current_migration in self.migration_sequence:
+            return self.migration_sequence.get(current_migration)
+        else:
+            return None
+
+    ## FHIR MANAGEMENT LOGIC
+    def get_latest_applied_migration_from_fhir(self) -> str:
+        """Retrieve the latest applied migration migration id from FHIR."""
+        # Logic to retrieve the latest applied migration number from FHIR
+        basic = MigrationManager.get_resource()
+        basic = first_in_bundle(basic)
+
+        if basic is None:
+            self.create_applied_migration_manager()
```

Comment:
i wouldn't create it, until we've run a successful migration.  just return None as you are doing here.